tests: add test case for kata agent API configuration

wainersm opened the pull request:
Added a e2e test to check that when a pod is created with an agent API
config that blocked certain endpoints that requests to these are rejected.

Fixes #4676
Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

wainersm commented:
Hi! I'm labeling 'rfc' this PR because so far my goal is to collect feedback.

I'm introducing a method (`cp_to_guest_img()`) to allow us to copy files to the guest image. Currently, on CCv0 branch we build the image with some files for testing included, but we likely must stop doing that when merge the branches. In a particular chat with @fidencio he suggested the use of volumes (IIUC pod volumes) but I should have the file inside the guest at the time the agent starts. Did you like that approach?

I noticed that when I enable the "CC" handler in containerd's configuration and I *don't* enable image offloading on kata side then it fails to start the container. Is it expected behavior? i.e. once you add the handler, you are completely disabling the image pulls from containerd (delegated to the kata agent) so it is mandatory to enable image offload on kata?

My last remark is that it seems we don't necessarily need to test agent API configuration with CC. If that is the case maybe I should try to merge that test case straight to the main branch. Or am I missing something?

amshinde commented on integration/containerd/confidential/agent_api.bats:
```diff
@@ -0,0 +1,66 @@
+	# operation should fail
+	! crictl exec "$container_id" cat /proc/cmdline
+
+	# TODO: grep the logs to be sure if failed because of the blocked endpoint.
```
@wainersm Are you planning to add this check for logs ?

wainersm replied:
Hi @amshinde , yes, I will lift the rfc label and send a new version of this PR where I am going to introduce that check.

wainersm commented:
Updates to this pull request:
 * added code to save/restore the guest image.

wainersm commented:
Updates to this pull request:
  * Added a checker on logs to ensure the cause of the failure was the API blocked
  * Added a checker for the default case: no passing a configuration file, the exec API should work
  * Created a shared library (asserts.sh) to hold all assert functions
  * using those asserts in both agent_image and agent_api tests

stevenhorsman commented on integration/containerd/confidential/agent_api.bats:
```diff
@@ -0,0 +1,83 @@
+load "${BATS_TEST_DIRNAME}/../../../lib/common.bash"
+
+setup() {
+	start_date=$(date +"%Y-%m-%d %H:%M:%S")
+	sandbox_name="kata-cc-busybox-sandbox"
+	pod_config="${FIXTURES_DIR}/pod-config.yaml"
+	pod_id=""
+
+	echo "Delete any existing ${sandbox_name} pod"
+	crictl_delete_cc_pod_if_exists "$sandbox_name"
+
+	echo "Prepare containerd for Confidential Container"
+	SAVED_CONTAINERD_CONF_FILE="/etc/containerd/config.toml.$$"
+	configure_cc_containerd "$SAVED_CONTAINERD_CONF_FILE"
+
+	echo "Reconfigure Kata Containers"
+	clear_kernel_params
+	switch_image_service_offload on
```
As this code matches the `agent_image.bats` I wonder if there is merit in moving it to a shared library to avoid the duplication?

wainersm replied:
Yeah, It does make sense to remove that duplication. I wish it was Python (or any OO language) where we could use heritage. Mind if I send it in a follow up PR?

stevenhorsman reviewed (approved):
Hi Wainer, sorry I missed this PR. I think it looks good and having a method to add files to the guest image is going to be really helpful and something we should switch to using in `ccv0.sh`.
With respect to your question about where to test agent API - I think you are correct that the function went into `main` and your test case would be valid in a non CC environment.

wainersm commented:
Update this PR with what I think the last missing bits: added handler for initrd in `cp_to_guest_img()` (https://github.com/kata-containers/tests/pull/4677/commits/40d7f5d0a9428bc2ef2bc455bcb83d13ff1dbfe0). Sending in a separate commit just to ease the reviews but I can squash into b306e7a726d6f60322236ae34faf38f6f6228177 before the merge.

wainersm commented:
Hi @stevenhorsman ,

> Hi Wainer, sorry I missed this PR. I think it looks good and having a method to add files to the guest image is going to be really helpful and something we should switch to using in `ccv0.sh`. With respect to your question about where to test agent API - I think you are correct that the function went into `main` and your test case would be valid in a non CC environment.

Can we adapt `ccv0.sh` after merging this one? Thinking aloud...maybe it can be done alongside the merge to `main`. Once we have https://github.com/kata-containers/tests/pull/4675 merged and ported to `CCv0`, we can work on preparing the osbuilder (**) which, among other changes, it will need to remove the signature and config files being added in the image at the build time.

** it seems we don't have an issue to track that work. At least the closest that I found was about removing skopeo/umoci (https://github.com/kata-containers/kata-containers/issues/3970)

---

Regarding where to merge this PR...I think it is too late to merge on `main` as it is too tied with the code in `CCv0`. Mind if I port it alongside the other CC tests when the time comes for https://github.com/kata-containers/tests/issues/4628?

wainersm commented:
/test

stevenhorsman commented:
Hey Wainer, sorry for the lack of clarity, my comments weren't supposed to be merge stoppers, just something to consider.

> Can we adapt `ccv0.sh` after merging this one? Thinking aloud...maybe it can be done alongside the merge to `main`. Once we have #4675 merged and ported to `CCv0`, we can work on preparing the osbuilder (**) which, among other changes, it will need to remove the signature and config files being added in the image at the build time.

Agreed, I meant this is great code we should try and switch to using in `ccv0.sh` in a separate PR after this as refactoring to remove the hard-coded policy files from the dev code will be great. I don't think that should be in this PR though :)

> Regarding where to merge this PR...I think it is too late to merge on `main` as it is too tied with the code in `CCv0`. Mind if I port it alongside the other CC tests when the time comes for #4628?

Yep, that's fine by me. My comment was indented to mean that the scenario would be valid with the current `main` dev, but can be ported as part of the exisiting plans.

stevenhorsman commented:
Hey @wainersm - as you mentioned that it was missing, I've created https://github.com/kata-containers/kata-containers/issues/4111 to cover using the new `cp_to_guest_img` to inject the image container image signature verification files into the guest image. Please take a look and add an info & fix my mistakes. This might be something that someone at IBM can do soon, but no promises yet until we've done some planning :)
